GH-2268: Add RedisHeaders.MESSAGE_SOURCE header

artembilan opened the pull request:
Resolves: /spring-projects/spring-integration#2268

* To indicate the source the Redis message in the `RedisInboundChannelAdapter`
populate the `RedisHeaders.MESSAGE_SOURCE` header to the messages to produce
* Fix the `SimpleMessageConverter` to populate the provided `MessageHeaders`
to the message to produce

artembilan committed:
GH-2268: Add RedisHeaders.MESSAGE_SOURCE header

Resolves: /spring-projects/spring-integration#2268

* To indicate the source the Redis message in the `RedisInboundChannelAdapter`
populate the `RedisHeaders.MESSAGE_SOURCE` header to the messages to produce
* Fix the `SimpleMessageConverter` to populate the provided `MessageHeaders`
to the message to produce

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/support/converter/SimpleMessageConverter.java:
```diff
@@ -103,7 +105,14 @@ protected MessageBuilderFactory getMessageBuilderFactory() {
+						MessageBuilder.fromMessage(message)
+								.copyHeadersIfAbsent(headers)
+								.build();
+			}
```
Hmmm - this will impact every messaging gateway. I wonder if we could have the mapper return the builder (perhaps a `default` method)?

Or provide an overloaded method that takes extra headers.

Or create a new converter that's only used by Redis.

artembilan replied:
As you see I have switched `RedisInboundChannelAdapter` to use `org.springframework.messaging.converter.SimpleMessageConverter` instead.

But I agree, I'm looking now to something "not very impacting"...

garyrussell commented on spring-integration-redis/src/main/java/org/springframework/integration/redis/inbound/RedisInboundChannelAdapter.java:
```diff
@@ -78,6 +82,12 @@ public void setMessageConverter(MessageConverter messageConverter) {
+	/**
+	 * Specify an {@link Executor} used for running the message listeners when messages are received.
+	 * @param taskExecutor the Executor to use for listener container.
+	 * @since 4.3.13
```
Does this __really__ need to be back-ported? There's a work around (header enricher).

artembilan replied:
Well, that's just a JavaDocs from the previous commit.

We indeed provide this `MESSAGE_SOURCE` feature only for `5.0`.

At the same time I'm not sure how you are going to workaround this header when we already step out of this `RedisInboundChannelAdapter`

garyrussell commented on src/reference/asciidoc/redis.adoc:
```diff
@@ -103,7 +103,7 @@ Instead those Messages are passed through Redis allowing you to rely on its supp
 ==== Redis Inbound Channel Adapter
 
-The Redis-based Inbound Channel Adapter adapts incoming Redis messages into Spring Integration Messages in the same way as other inbound adapters.
+The Redis-based Inbound Channel Adapter (`RedisInboundChannelAdapter`) adapts incoming Redis messages into Spring Integration Messages in the same way as other inbound adapters.
```
`spring-messaging` messages?

artembilan committed:
* Add `toMessage(T object, @Nullable Map<String, Object> headers)`
to the `InboundMessageMapper` to propagate additional header
to the message to create
* Rework all the out-of-the-box `InboundMessageMapper` implementations
to properly propagate additional headers via `toMessage()`
from the `SimpleMessageConverter`
* Provide optimizations in the `InboundMessageMapper` implementations
do not re-create messages
* Refactor `MutableMessage.toString()` to align with the `GenericMessage`

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/gateway/MessagingGatewaySupport.java:
```diff
@@ -792,11 +793,14 @@ void setMessageBuilderFactory(MessageBuilderFactory messageBuilderFactory) {
-			return (object != null) ? this.messageBuilderFactory.withPayload(object).build() : null;
+
+			return object != null
+					? this.messageBuilderFactory.withPayload(object).copyHeadersIfAbsent(headers).build()
```
We should add `@Nullable` to `copyHeadersIfAbsent` (it's present on the Spring Messaging MB).

artembilan replied:
Is it related to this PR ?

There are too much those Kotlin "love" in SF...

garyrussell replied:
Yes, it's related because you now unconditionally `copyHeadersIfAbsent(headers)` - I had to go and check the code to see if it's protected. `@Nullable` documents it.

artembilan replied:
??? That was always unconditional because there is `if (headersToCopy != null) {`.

I don't understand your concern regarding this `@Nullable`. 
This is something new in SF `5.0` and we have lived without it for a while already.

What is the point to jumt into this subject now?
Thanks

garyrussell replied:
? This cHIA () is new code in this commit and the arg is @Nullable.

artembilan replied:
How is that new? 
I just use here existing `MessageBuilder` API.
What am I missing?

Right, I added a new `headers` arg to the `toMessage()` method and just followed its interface contract with adding `@Nullable`. And this is definitely new code. But why do I have to go from here everywhere to fix all the possible `@Nullable`?

Cant't we address it on demand at least?

garyrussell replied:
You don't have to "go everywhere". Since you can send null here I had to check if it was safe. So this is indeed "on demand". This discussion is more effort than it would have taken ro fix it.

artembilan committed:
* Add more `@Nullable` to method arguments
* Increase latch wait timeout in the `EndpointParserTests`
* Address `redis.adoc` PR comment

artembilan commented:
Pushed.

Thank you for the patience, Gary, and for the clean explanation what you want to see.
Although I still don't see a gain of all those `@Nullable`s 😄

garyrussell commented:
Merged as 7ca20e53f78688f46617775b0d45ba8266e817e1
